spv/verifypayment.go

```diff
@@ -4,176 +4,31 @@ import (
-	}
-	// parse initial tx, fail fast if it isn't a valid tx.
-	tx, err := bt.NewTxFromString(initialPayment.RawTx)
+	ancestry, err := NewAncestryFromBytes(ancestors)
 	if err != nil {
 		return nil, err
 	}
-
-	// verify tx fees
-	if vOpt.fees {
-		if err := v.verifyFees(initialPayment, tx, vOpt); err != nil {
-			return nil, err
-		}
-	}
-	// if we are validating proofs or scripts carry out the validation.
-	if vOpt.requiresEnvelope() {
-		// The tip tx is the transaction we're trying to verify, and it should not have a supplied
-		// Merkle Proof.
-		if initialPayment.IsAnchored() {
-			return nil, ErrTipTxConfirmed
-		}
-		if err := v.verifyTxs(ctx, initialPayment, vOpt); err != nil {
-			return nil, err
-		}
-	}
-	return tx, nil
-}
-
-// verifyFees takes the initial payment and iterates the immediate parents in order to gather
-// the satoshis used for each input of the initialPayment tx.
-//
-// If there are no parents the method will fail, also, if there are no fees the method will fail.
-func (v *verifier) verifyFees(initialPayment *Envelope, tx *bt.Tx, opts *verifyOptions) error {
-	if !initialPayment.HasParents() {
-		return ErrCannotCalculateFeePaid
-	}
-	if opts.feeQuote == nil {
-		return ErrNoFeeQuoteSupplied
-	}
-	for _, input := range tx.Inputs {
-		parent, err := initialPayment.ParentTx(input.PreviousTxIDStr())
-		if err != nil {
-			return errors.Wrapf(err, "tx %s failed to get parent tx", tx.TxID())
-		}
-		out := parent.OutputIdx(int(input.PreviousTxOutIndex))
-		if out == nil {
-			return ErrMissingOutput
-		}
-		input.PreviousTxSatoshis = out.Satoshis
-	}
-	ok, err := tx.IsFeePaidEnough(opts.feeQuote)
-	if err != nil {
-		return err
-	}
-	if !ok {
-		return ErrFeePaidNotEnough
-	}
-	return nil
-}
-
-func (v *verifier) verifyTxs(ctx context.Context, payment *Envelope, opts *verifyOptions) error {
-	// If at the beginning or middle of the tx chain and tx is unconfirmed, fail and error.
-	if opts.proofs && !payment.IsAnchored() && !payment.HasParents() {
-		return errors.Wrapf(ErrNoConfirmedTransaction, "tx %s has no confirmed/anchored tx", payment.TxID)
-	}
-
-	// Recurse back to the anchor transactions of the transaction chain and verify forward towards
-	// the tip transaction. This way, we check that the first transactions in the chain are anchored
-	// to the blockchain through a valid Merkle Proof.
-	for parentTxID, parent := range payment.Parents {
-		if parent.TxID == "" {
-			parent.TxID = parentTxID
-		}
-		if err := v.verifyTxs(ctx, parent, opts); err != nil {
-			return err
-		}
-	}
-
-	// If a Merkle Proof is provided, assume we are at the anchor/beginning of the tx chain.
-	// Verify and return the result.
-	if payment.IsAnchored() || !payment.HasParents() {
-		if opts.proofs {
-			return v.verifyTxAnchor(ctx, payment)
-		}
-		return nil
-	}
-
-	tx, err := bt.NewTxFromString(payment.RawTx)
+	paymentTx, err := bt.NewTxFromBytes(pTx)
```

Comment:
Is there a reason why the param for `pTx` isn't `*bt.Tx` and instead is `[]byte`?

I'm just thinking, there's a chance that a user might already have it as a `*bt.Tx`, and will need to perform the marshaling to bytes, just for the tx to be marshalled back again.

I don't think we hit this problem in reverse, I could be wrong tho

Reply:
I was imagining the payment coming in as binary - but there's no reason it has to be like this - happy to update.

Reply:
I'm thinking about it coming in as a binary, someone converting it to a bt.Tx first for whatever reason, passing it to a function, then passing tx.Bytes() into this, just to have NewFromBytes called again.

It might a non-issue this, what do you reckon?